source/reduce/reduction_pass.cpp

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+
+#include "reduction_pass.h"
+
+#include "source/opt/build_module.h"
+
+namespace spvtools {
+namespace reduce {
+
+std::vector<uint32_t> ReductionPass::ApplyReduction(
+    const std::vector<uint32_t>& binary) {
+  std::unique_ptr<opt::IRContext> context =
+      BuildModule(target_env_, consumer_, binary.data(), binary.size());
```

Comment:
Is there a reason you want to build the module every time you apply a reduction pass?  If this is done a lot, it could end up being time consuming.  At the very least add a comment explaining why you rebuild the module and context every time.

Reply:
I'll add this comment:

  // We represent modules as binaries because (a) attempts at reduction need to end up in binary form to be passed on
  // to SPIR-V-consuming tools, and (b) when we apply a reduction step we need to do it on a fresh version of the
  // module as if the reduction step proves to be uninteresting we need to backtrack; re-parsing from binary provides
  // a very clean way of cloning the module.